File: source/downloads/code/LeetCode/Python/Generate-Parentheses.py
class Solution(object):
    def getParenthesis(self, l, r, tmp, ret):
        # 有括号比左括号多
        if r < l:
            return
        if l ==0 and r == 0:
            ret.append(tmp)
        
        if l > 0:
            self.getParenthesis(l-1,r,tmp+'(',ret)
        if r > 0:
            self.getParenthesis(l,r-1,tmp+')',ret)


    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        
        ret = []
        if n == 0:
            return ret

        self.getParenthesis(n,n,"",ret)
        return ret

File: downloads/code/LeetCode/Java/Generate-Parentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ret = new LinkedList<String>();    
        dfs(ret,"",0,0,n);
        return ret;
    }

    public void dfs(List<String> ret, String tmp, int l, int r, int max){
            if(tmp.length() == 2*max){
                ret.add(tmp);
                return;
            }
            if(l<max)
                dfs(ret, tmp+'(', l+1, r, max);
            if(r<l)
                dfs(ret, tmp+')', l, r+1, max);
    }
}
